File: backend/app/models/loyalty.py
from pydantic import BaseModel


class NewLoyaltySnapshot(BaseModel):
    date: str
    customer_retention_rate: float       # %
    repeat_purchase_rate: float          # %
    avg_clv: float                        # currency
    partnership_renewal_rate: float      # %
    referral_rate: float                  # %


class LoyaltySnapshot(NewLoyaltySnapshot):
    row_index: int

File: frontend/src/api/loyalty.ts
import { api } from "./client";

export interface LoyaltySnapshot {
  row_index: number;
  date: string;
  customer_retention_rate: number;
  repeat_purchase_rate: number;
  avg_clv: number;
  partnership_renewal_rate: number;
  referral_rate: number;
}

export interface NewLoyaltySnapshot {
  date: string;
  customer_retention_rate: number;
  repeat_purchase_rate: number;
  avg_clv: number;
  partnership_renewal_rate: number;
  referral_rate: number;
}

export async function getLoyaltySnapshots(): Promise<LoyaltySnapshot[]> {
  const { data } = await api.get<LoyaltySnapshot[]>("/loyalty/snapshots");
  return data;
}

export async function addLoyaltySnapshot(snapshot: NewLoyaltySnapshot): Promise<void> {
  await api.post("/loyalty/snapshots", snapshot);
}

export async function deleteLoyaltySnapshot(rowIndex: number): Promise<void> {
  await api.delete(`/loyalty/snapshots/${rowIndex}`);
}
